lockbox-ios/Presenter/ItemListPresenter.swift

```diff
@@ -38,17 +54,33 @@ class ItemListPresenter {
     }
 
     func onViewReady() {
-        let listDriver = self.dataStore.onItemList
-                .do(onNext: { items in
-                    if items.isEmpty {
+        let listDriver = Observable.combineLatest(self.dataStore.onItemList, self.filterTextSubject.asObservable())
+                .map { (latest: ([Item], String)) -> ItemsText in
+                    return ItemsText(items: latest.0, text: latest.1)
+                }
+                .distinctUntilChanged()
+                .do(onNext: { latest in
+                    if latest.items.isEmpty {
                         self.view?.displayEmptyStateMessaging()
                     } else {
                         self.view?.hideEmptyStateMessaging()
                     }
                 })
-                .filter { items in
-                    return !items.isEmpty
+                .filter { latest in
+                    return !latest.items.isEmpty
                 }
+                .map { (latest: ItemsText) -> [Item] in
+                    if latest.text.isEmpty {
+                        return latest.items
+                    }
+
+                    let latestText = latest.text.lowercased()
+                    return latest.items.filter { item -> Bool in
+                        return item.entry.username?.lowercased().contains(latestText) ?? false
```

Comment:
Be really wary about this — if your text is user-enterable, then you can't assume converting to lowercase and comparing will be correct.

Instead, use:
```swift
a.caseInsensitiveCompare(b) == .orderedSame
```

Reply:
I'd also group the optional unwrapping together: 

```swift
guard let itemText = item.entry.username ?? item.origins.first ?? item.title else {
    return false
}
return itemText.caseInsensitiveCompare(latest.text) == .orderedSame
```

Reply:
`.caseInsensitiveCompare()` does a complete comparison; we should return matches even if the entered text only matches a substring of the username or title or origin.

Reply:
Gah! I misread. 

Nevertheless: we shouldn't be rolling our own case insensitive string ops. They aren't locale aware; there are alphabets that don't have a 1-1 mapping between upper and lower cases. 

Strings are weird. Rely on the OS (via [ICU][1]) to do the work for you.

```swift
guard let itemText = item.entry.username ?? item.origins.first ?? item.title else {
    return false
}
return itemText.localizedCaseInsensitiveContains(latest.text)
```

[1]: http://site.icu-project.org/